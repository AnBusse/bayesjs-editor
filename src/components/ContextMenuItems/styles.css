.context-menu {
  position: fixed;
  border: solid 1px #cbcbcb;
  box-shadow: 4px 4px 6px -3px #666;
  margin: 0;
  padding: 0;
  list-style: none;
  top: 0;
  left: 0;
  z-index: 1;
}

.context-menu-item {
  user-select: none;
  cursor: default;
  padding: 5px 50px 5px 25px;
  background-color: #fff;
  white-space: nowrap;
}

.context-menu-item--disabled {
  cursor: not-allowed;
  color: #a5a1a1;
}

.context-menu-item:hover {
  background-color: #e5f3fb;
}

.context-menu-item--disabled:hover {
  background-color: #fff;
}
