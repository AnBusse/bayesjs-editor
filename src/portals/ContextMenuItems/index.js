import ReactDOM from 'react-dom';

const contextItemsElement = document.getElementById('context-items');

const ContextMenuItems = ({ children }) => ReactDOM.createPortal(
  children,
  contextItemsElement,
);

export default ContextMenuItems;
