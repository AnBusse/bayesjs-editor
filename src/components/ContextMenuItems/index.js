import React, { useState } from 'react';
import ContextMenuItemsPortal from 'portals/ContextMenuItems';
import { ContextMenu, MenuItem } from 'react-contextmenu';
import PropTypes from 'prop-types';
import {
  merge,
  complement,
  isNil,
  is,
  join,
  concat,
} from 'ramda';
import { contextMenuItemPropTypes } from 'models';
import { subtractPositions } from 'utils/position';
import { getComponentTestId } from 'utils/test-utils';
import styles from './styles.css';

const isNotNil = complement(isNil);
const isFunction = is(Function);

const onShowHandler = setMousePosition => ({ detail: { position, data: { target } } }) => {
  const targetPosition = target.getBoundingClientRect();
  const finalPosition = subtractPositions(position, targetPosition);

  setMousePosition(finalPosition);
};

const onClickHandler = ({ onClick }, mousePosition) => (event, ...args) =>
  onClick(merge(event, { mousePosition }), ...args);

const isItemDisabled = ({ disabled }, data) => {
  if (isNotNil(disabled)) {
    return isFunction(disabled) ? disabled(data) : disabled;
  }

  return false;
};

const getItemClasses = isDisabled =>
  join(' ', [
    styles.contextMenuItem,
    ...(isDisabled ? [styles.contextMenuItemDisabled] : []),
  ]);

const ContextMenuItems = ({
  id,
  type,
  items,
  data,
}) => {
  const [mousePosition, setMousePosition] = useState();

  return (
    <ContextMenuItemsPortal>
      <ContextMenu
        className={styles.contextMenu}
        id={concat(id, type)}
        onShow={onShowHandler(setMousePosition)}
      >
        {items.map(({
          key,
          text,
          style,
          ...itemProps
        }) => {
          const isDisabled = isItemDisabled(itemProps, data);
          const classNames = getItemClasses(isDisabled);

          return (
            <MenuItem
              key={key}
              onClick={onClickHandler(itemProps, mousePosition)}
              data={data}
              disabled={isDisabled}
            >
              <div
                className={classNames}
                style={style}
                data-testid={getComponentTestId('ContextMenuItems', key)}
                children={text}
              />
            </MenuItem>
          );
        })}
      </ContextMenu>
    </ContextMenuItemsPortal>
  );
};

ContextMenuItems.propTypes = {
  id: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  items: PropTypes.arrayOf(contextMenuItemPropTypes).isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  data: PropTypes.any,
};

ContextMenuItems.defaultProps = {
  data: null,
};

export default ContextMenuItems;
