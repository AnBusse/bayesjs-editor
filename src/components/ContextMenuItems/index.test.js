import { shallow } from 'enzyme';
import React from 'react';
import { ContextMenu } from 'react-contextmenu';
import ContextMenuItems from './index';

const defaultProps = {
  id: 'id',
  type: 'type',
  items: [
    {
      key: '1',
      text: 'Item 1',
      disabled: () => false,
      onClick: jest.fn(),
    },
    {
      key: '2',
      text: 'Item 2',
      disabled: true,
      onClick: jest.fn(),
    },
    {
      key: '3',
      text: 'Item 3',
      onClick: jest.fn(),
    },
  ],
  data: null,
};

const shallowComponent = () =>
  shallow(<ContextMenuItems {...defaultProps} />);

describe('ContextMenuItems Component', () => {
  it('matches snapshot', () => {
    const component = shallowComponent();

    expect(component).toMatchSnapshot();
  });

  it('has id and type as prop id', () => {
    const component = shallowComponent();

    expect(component.find(ContextMenu).prop('id')).toBe(`${defaultProps.id}${defaultProps.type}`);
  });
});
