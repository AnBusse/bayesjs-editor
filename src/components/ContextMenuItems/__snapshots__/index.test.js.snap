// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextMenuItems Component matches snapshot 1`] = `
<ContextMenuItems>
  <ContextMenu
    className="contextMenu"
    data={Object {}}
    hideOnLeave={false}
    id="idtype"
    onHide={[Function]}
    onMouseLeave={[Function]}
    onShow={[Function]}
    rtl={false}
    style={Object {}}
  >
    <MenuItem
      attributes={Object {}}
      data={null}
      disabled={false}
      divider={false}
      key="1"
      onClick={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      preventClose={false}
      selected={false}
    >
      <div
        className="contextMenuItem"
        data-testid="context-menu-items__1"
      >
        Item 1
      </div>
    </MenuItem>
    <MenuItem
      attributes={Object {}}
      data={null}
      disabled={true}
      divider={false}
      key="2"
      onClick={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      preventClose={false}
      selected={false}
    >
      <div
        className="contextMenuItem contextMenuItemDisabled"
        data-testid="context-menu-items__2"
      >
        Item 2
      </div>
    </MenuItem>
    <MenuItem
      attributes={Object {}}
      data={null}
      disabled={false}
      divider={false}
      key="3"
      onClick={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      preventClose={false}
      selected={false}
    >
      <div
        className="contextMenuItem"
        data-testid="context-menu-items__3"
      >
        Item 3
      </div>
    </MenuItem>
  </ContextMenu>
</ContextMenuItems>
`;
